### ISSUE TITLE
[Feature request] Admin rights handling when swapping DLSS in Xbox app games

### ISSUE THREAD

**Issue by revanmj** [beeradmoore/dlss-swapper#91]
**Is your feature request related to a problem? Please describe.**
As discussed in #2, it seems that in some cases Xbox app game folders require admin rights for file operations inside them and DLSS Swapper currently doesn't handle it well (it just fails with generic error suggesting repairing game files, but works fine if whole app is launched as admin).

**Describe the solution you'd like**
Maybe in case of a fail retry swapping with admin rights just for Xbox app games or add additional error dialog with such option?

**Additional context**
My guess is that on PCs updated to Win10/11 from older versions, partitions or permissions may have been handled differently than on clean installs of those systems which leads to this (as I don't have this problem on my laptop with clean install of Win11, only on PC which was not formatted since Vista, only upgraded).

**Comment by beeradmoore** [comment]
Bringing across my comment from the other thread.

==========
Interesting... 

For me on both of my systems (neither of which I have done funky things with WindowsApps folder) I can write to the game folder while using my normal accounts. I can't view the WindowsApps folder itself, so if I am in a game folder and attempt to go up a directory it says no.

For reference of what I have going on here - one system is Win11 where I am logged in with my MS account. The other system is Win10 where I am logged in as a local account. Both systems I am logged into the same MS account for Xbox app.

This new feature works by using the .GamingRoot files which are stored on the root of every drive. It contains the location of your Xbox games (what is the right word for this, Xbox app games?) are installed on that drive. It then lists these directories and compares against a list of every package installed on your computer to say. It then keeps a reference to the game location in WindowsApp folder which as I said above, the individual folders (at least these ones) are fully writable to me. The files added here also instantly add to the other folder (C:\XboxGames\ for me)

I am in Xbox app Insider and I have used the [Advanced Installation and Management Features](https://news.xbox.com/en-us/2021/11/18/advanced-installation-and-management-features-available-for-insiders-through-the-xbox-app-for-windows/) for the Xbox app, but the games I had tested with were not "advanced", they just let me install the game location.

My test game was The Ascent from Game Pass.
==========

Some additional info to add is both PCs I have I did install that particular OS from scratch. I will look into checking permission and prompting for admin access where required though.

**Comment by revanmj** [comment]
Regarding the folder, I used D:\Program Files\Xbox Games\ in case it could also matter (I have 3 physical SSDs with one partition on each)

**Comment by beeradmoore** [comment]
It could be program files thing. I tried to move my folder there and it asked me for permission to do it. It then would complain about trying to make the folder.

If I manually make the folder in there I can select it from within the Xbox app but it doesn't seem to change from what I had 🤷‍♂️

**Comment by beeradmoore** [comment]
I was able to set a game to use this folder. I created `H:\Program Filez\xboxgames\` (note the z), I then made `xboxgames` in my normal `Program Files` directory and then edited the `.GamingRoot` file to have `s` not `z`. Rebooted, and now I have ascent installing to this folder. Hopefully I can replicate the issue once this installs.

**Comment by beeradmoore** [comment]
I don't adjust permissions on my system for those folders as every time I do I mess it up :P 

I can't see how to copy the specific file as admin, but I can handle the error better for these specific errors.

![image](https://user-images.githubusercontent.com/904737/194831435-bf89daa4-d9d9-4c61-bfa6-32c1f531de22.png)

**Comment by beeradmoore** [comment]
So I can make UAC prompt by telling DLSS Swapper to copy the files with command line as admin. Trouble is it will prompt at minimum twice.

It first does it to make a backup of the dll next to the original. It then does it a second time to copy the updated DLL over the original. But it is also twice for each dll found, so if a game has the dll sitting around twice for some dumb reason it will prompt UAC 4 times.

May be best to have the above prompt but also try "restart dlss swapper as admin" button.

**Comment by revanmj** [comment]
Yeah, with such number of requests, restart as admin would be better

**Comment by beeradmoore** [comment]
I unfortunately could not get the button to re-launch as admin. But at least the above warning should help people. You can change the main exe to always run as admin, so you get a UAC prompt on launch just once, but adding those notes are not going into the app at this time. 

These changes have been committed and will be in the next release.

**Comment by revanmj** [comment]
> You can change the main exe to always run as admin

How? I tired finding exe in a folder and setting this on its Compatibility tab, but when launched from Start menu it still launches normally without UAC. Even directly launching via exe doesn't trigger UAC with this enabled. Seems that it doesn't work for packaged apps :/

**Comment by revanmj** [comment]
Also, I've read a little on the subject out of curiosity, and to do this with only one request in-app, you would need a separate exe that just does all the file operations (or ability to run existing one as a separate process that will do the same).

**Comment by beeradmoore** [comment]
Oh, my mistake. It may have been when I was using local builds. I was playing around with it and would have it say "is administrator: true/false" on launch. But running even in release mode may not have been unpackaged so it let me do that 🤔

One of the things I was trying to re-launch as admin was to launch cmd.exe as its own exe to then launch DLSS Swapper again after it had closed but that didn't work. As for the second process I believe to only have one prompt I'd need to keep that exe running and then communicate between the two otherwise I think it prompts every time it starts up. That communication layer complicates things a little. You also would want to make sure those two processes only communicate with each other, otherwise we'd have an exe just running as admin that any malicious program could tap into to start copying files.

I'll re-open this to make sure the issue is not forgotten about.

**Comment by beeradmoore** [comment]
Also regarding packaged apps. Now that WinUI3 supports it I'd like to move GitHub releases to be unpackaged to be installed with an installer and if possible, have another distribution which is then completely portable to satisfy #67. I'll need to tweak some of the places we download things to (and in future cache images) but that shouldn't be too bad.

It also means I don't have to keep renewing the certificate every year that is distributed in order to install the app from GitHub instead of Windows Store.

**Comment by beeradmoore** [comment]
This should now be resolved in [DLSS Swapper v1.1](https://github.com/beeradmoore/dlss-swapper/releases/tag/v1.1)

I am closing this issue now, feel free to continue discussing in this thread. If you are still experiecing problems please open a new issue

**Comment by revanmj** [comment]
Well, it shows an error banner saying that it couldn't swap the file due to lack of admin permission and that's it. So, solution is still ticking "run as admin" option in compatibility tab of the app's exe file properties and always run it as admin.

**Comment by beeradmoore** [comment]
Can you let me know what path DLSS Swapper.exe is running from, and what folder you are trying to swap to?

**Comment by revanmj** [comment]
DLSS Swapper is in C:\Program Files\DLSS Swapper

OK, here may be an issue. 1.1 detects MS Store games paths always in C:\Program Files\WindowsApps\...

Even though Xbox app shows proper paths for them (like I:\Program Files\Xbox\NINJA GAIDEN 2 Black\)

**Comment by beeradmoore** [comment]
That's indeed odd, especially as is a new game. If it were installed 3 years ago I'd assume it's a problem of when they changed how install paths work

**Comment by beeradmoore** [comment]
When I can I'll install DCS and try replicate
